arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceOptionsViewController.swift

```diff
@@ -0,0 +1,271 @@
+// limitations under the License.
+
+import UIKit
+import ArcGIS
+
+protocol ConfigureSubnetworkTraceOptionsViewControllerDelegate: AnyObject {
+    func onDismiss(_ controller: ConfigureSubnetworkTraceOptionsViewController, didAddContidionExpression expression: AGSUtilityTraceConditionalExpression)
+}
+
+class ConfigureSubnetworkTraceOptionsViewController: UITableViewController {
+    // MARK: Storyboard views
+    
+    /// The cell for attribute options.
+    @IBOutlet var attributesCell: UITableViewCell!
+    /// The cell for comparison operator options.
+    @IBOutlet var comparisonCell: UITableViewCell!
+    /// The cell for value to compare with.
+    @IBOutlet var valueCell: UITableViewCell!
+    /// A button to add the conditional expression to the trace configuration.
+    @IBOutlet var addBarButtonItem: UIBarButtonItem!
+    
+    // MARK: Properties
+    
+    /// A delegate to notify other view controllers.
+    weak var delegate: ConfigureSubnetworkTraceOptionsViewControllerDelegate?
+    
+    /// An array of possible network attributes.
+    var possibleAttributes: [AGSUtilityNetworkAttribute]!
+    /// An array of `AGSUtilityAttributeComparisonOperator` and their description string pairs.
+    var attributeComparisonOperators: KeyValuePairs<AGSUtilityAttributeComparisonOperator, String>!
+    
+    /// The attribute selected by the user.
+    var selectedAttribute: AGSUtilityNetworkAttribute? {
+        didSet {
+            // Reset the selected value.
+            attributesCell.detailTextLabel?.text = selectedAttribute?.name
+            selectedValue = nil
+            updateCellStates()
+        }
+    }
+    /// The comparison selected by the user.
+    var selectedComparison: AGSUtilityAttributeComparisonOperator? {
+        didSet {
+            if let index = attributeComparisonOperators.firstIndex(where: { $0.0 == selectedComparison }) {
+                comparisonCell.detailTextLabel?.text = attributeComparisonOperators[index].1
+            } else {
+                comparisonCell.detailTextLabel?.text = nil
+            }
+            updateCellStates()
+        }
+    }
+    /// The value selected by the user.
+    var selectedValue: Any? {
+        didSet {
+            addBarButtonItem.isEnabled = selectedValue != nil
+        }
+    }
+    
+    /// A reference to the observer that detects if text field input is empty.
+    var emptyStringObserver: Any!
+    
+    // MARK: Actions
+    
+    @IBAction func addConditionBarButtonItemTapped(_ sender: UIBarButtonItem) {
+        if let attribute = selectedAttribute, let comparison = selectedComparison, let value = selectedValue {
+            let convertedValue: Any
+            
+            if let codedValue = value as? AGSCodedValue, attribute.domain is AGSCodedValueDomain {
+                // The value is a coded value.
+                convertedValue = convertToDataType(value: codedValue.code!, dataType: attribute.dataType)
```

Comment:
> Also, for attributes with predefined coded values, my implementation displays the name of the coded values (A, AB, etc). Not sure where the confusion is here.

I made changes to the `convertToDataType` function, and it no longer give results other than numbers and booleans. In your original implementation it mostly default to return the coded value type itself, while now it becomes a Swift value type. I didn't scrutinize the original design - is converting coded value needed? If not I can make changes to this line.

Reply:
Edit: it seems you explicitly convert a coded value back to its name by finding it from all the values in its domain. I can change to this approach.

Reply:
Yes, please change it. Thanks!

Reply:
It should be good now!